### PR TITLE
[Bug: #154207101] Update user schema 

### PR DESCRIPTION
**Pull request by fisayomi**
#### What does this PR do?
- Adds UpdateUser Mutation
- Fixes bugs in readme and user schema

#### Description of Task to be completed?
- Enable Users to update desired fields.

#### How should this be manually tested?
- Run the UpdateUser and CreateUser Mutations

#### Any background context you want to provide?
- An error message should be retuned if creation of a User fails. 

#### What are the relevant pivotal tracker stories?
[#154207101](https://www.pivotaltracker.com/story/show/154207101)

### CODE REVIEW

File: food_basket/schema/user.py [hunk 1]
```diff
@@ -111,7 +116,39 @@ def resolve_one_user(self, args, context, info):
 
         return None
 
+# Mutation for store update
+class UpdateUser(graphene.Mutation):
```
[the remark this answers is not in the document]
Comment (oreoluwade) [reply food_basket/schema/user.py]: 🍾

File: food_basket/tests/tests_001_user_schema.py [hunk 2]
```diff
@@ -163,3 +163,38 @@ def test_all_users_query(self):
...
+              '''
+        self.assertMatchSnapshot(client.execute(query))
+
+    # Test for store update failure mutation
+    def test_user_not_updated(self):
+        client = Client(schema)
+        query = '''
+                mutation updateUser{
+                  updateUser(
+                    storeId: 9,
```
Comment (oreoluwade) [inline food_basket/tests/tests_001_user_schema.py]: The error in the tests originates from here `storeId`

File: food_basket/tests/snapshots/snap_tests_001_user_schema.py [hunk 3]
```diff
@@ -93,3 +111,26 @@
...
+snapshots['UserTestCase::test_user_not_updated 1'] = {
+    'errors': [
+        {
+            'locations': [
+                {
+                    'column': 21,
+                    'line': 4
+                }
+            ],
+            'message': 'Unknown argument "storeId" on field "updateUser" of type "Mutations".'
```
[the remark this answers is not in the document]
Comment (oreoluwade) [reply food_basket/tests/snapshots/snap_tests_001_user_schema.py]: @fisayomi There's an error here. `storeId` must have been passed to the `updateUser` mutation.

File: food_basket/schema/user.py [hunk 4]
```diff
@@ -45,13 +46,17 @@ def mutate(self, cls, args, context, info):
...
-            user.set_password(password)
-            user.save()
             created_token = Token.objects.create(user=user)
             return CreateUser(user=user, token=created_token)
 
-        except:
-            return CreateUser()
+        except Error as e:
+            if 'username' in e.args[0]:
+                raise Error('Invalid Username')
```
Comment (oreoluwade) [inline food_basket/schema/user.py]: This line/case has not yet been tested

File: food_basket/schema/user.py [hunk 5]
```diff
@@ -45,13 +46,17 @@ def mutate(self, cls, args, context, info):
...
+            elif 'email' in e.args[0]:
+                raise Error("The email has been "
```
Comment (oreoluwade) [inline food_basket/schema/user.py]: This case has not been tested yet as well.

File: food_basket/schema/user.py [hunk 6]
```diff
@@ -111,7 +116,39 @@ def resolve_one_user(self, args, context, info):
...
+    def mutate(self, cls, args, context, info):
+        try:
+            user_id = args.get('user_id')
+            user = User.objects.get(id=user_id)
+
+            for arg in args:
+                if arg != 'user_id':
+                    setattr(user, arg, args.get(arg))
+                else:
+                    pass
```
Comment (oreoluwade) [inline food_basket/schema/user.py]: The `else` case here has also not been tested yet.

File: food_basket/schema/user.py [hunk 7]
```diff
@@ -111,7 +116,39 @@ def resolve_one_user(self, args, context, info):
...
+
+            user.save()
+            return UpdateUser(message="User successfully updated")
+
+        except Error:
```
Comment (oreoluwade) [inline food_basket/schema/user.py]: This case has also not been tested yet.

File: food_basket/tests/tests_001_user_schema.py [hunk 8]
```diff
@@ -163,3 +163,38 @@ def test_all_users_query(self):
...
+                        username
+                        email
+                    }
+                    message
+                  }
+                }
+              '''
+        self.assertMatchSnapshot(client.execute(query))
+
+    # Test for store update failure mutation
```
Comment (oreoluwade) [inline food_basket/tests/tests_001_user_schema.py]: I noticed a minor omission here. The comment should reference `user`, not `store`

File: food_basket/tests/tests_001_user_schema.py [hunk 9]
```diff
@@ -163,3 +163,38 @@ def test_all_users_query(self):
             }
         '''
         self.assertMatchSnapshot(client.execute(query))
+
```
Comment (oreoluwade) [inline food_basket/tests/tests_001_user_schema.py]: Please, kindly add a short comment here that describes what the test case handles.

Comment (fisayomi) [reply food_basket/tests/tests_001_user_schema.py]: @oreoluwade , I feel the def also describes what the test handles but will update as requested

Comment (oreoluwade) [reply food_basket/tests/tests_001_user_schema.py]: @fisayomi Yes, it definitely does. I only suggested it as an extra indicator for the tests, especially considering we've been doing that in other tests.

File: food_basket/schema/user.py [hunk 10]
```diff
@@ -111,7 +116,44 @@ def resolve_one_user(self, args, context, info):
...
+            user = User.objects.get(id=user_id)
+
+            for arg in args:
+                if arg != 'user_id':
+                    setattr(user, arg, args.get(arg))
+                else:
+                    pass
+
+            user.save()
+            return UpdateUser(message="User successfully updated")
```
Comment (oreoluwade) [inline food_basket/schema/user.py]: I think the user should also be passed as an argument to `UpdateUser` as well as the message:

`UpdateUser(user=user, message="User successfully updated")`

File: food_basket/schema/user.py (hunk 6 above)
Comment (fisayomi) [reply food_basket/schema/user.py]: I will request you try to work on this instance. @oreoluwade

File: food_basket/schema/user.py (hunk 10 above)
Comment (fisayomi) [reply food_basket/schema/user.py]: I think we do not need to return the user after update

File: food_basket/schema/user.py (hunk 6 above)
Comment (oreoluwade) [reply food_basket/schema/user.py]: @fisayomi Sure. When you are done with other comments you can handle, I would love to have a go at it. I don't want to start working and then have to `force-push`

File: food_basket/schema/user.py (hunk 10 above)
Comment (oreoluwade) [reply food_basket/schema/user.py]: Fair enough. We can easily get the user if we need any detail. 👍

File: .travis.yml [hunk 11]
```diff
@@ -13,7 +13,7 @@ install:
 before_script:
   - psql -c 'create database testdb;' -U postgres
 script:
-  - coverage run --source=food_basket manage.py test
+  - coverage run --source=food_basket --omit */snapshots/* manage.py test
```
Comment (oreoluwade) [inline .travis.yml]: Love this! 💯

File: food_basket/schema/user.py [hunk 12]
```diff
@@ -139,18 +139,10 @@ def mutate(self, cls, args, context, info):
...
                 if arg != 'user_id':
                     setattr(user, arg, args.get(arg))
-                else:
-                    pass
 
             user.save()
-            return UpdateUser(message="User successfully updated")
+            return UpdateUser(message="User successfully updated!")
 
-        except Error as e:
```
Comment (fisayomi) [inline food_basket/schema/user.py]: @oreoluwade  we need to check if the user is updating with existing values

**Comment by omorilewa** [comment]
Looks good. Great work.

File: food_basket/models/managers.py [hunk 13]
```diff
@@ -15,17 +13,3 @@ def create_user(self, email, password, **extra_fields):
         user.save(using=self._db)
         return user
 
-    # Commenting this block till we use the function
```
Comment (omorilewa) [inline food_basket/models/managers.py]: @oreoluwade I think this should be left in the meantime. Soon enough we would need to create a superuser.

File: food_basket/tests/tests_001_user_schema.py [hunk 14]
```diff
@@ -317,4 +280,4 @@ def test_update_user_mutation_wrong_value(self):
                   }
                 }
               '''
-        self.assertMatchSnapshot(client.execute(query))
```
Comment (andela-fojuri) [inline food_basket/tests/tests_001_user_schema.py]: @oreoluwade This is the part that we should return

Comment (andela-fojuri) [reply food_basket/tests/tests_001_user_schema.py]: lines 272- 303

**Pull request merged**